### PR TITLE
improve Esirkepov speed

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Speedup the Esirkepov current deposition by optimizing the number of deposition points around the particles.

- work with a virtual coordinate system (depending of the particle position)
- remove variable with fixed length loop's
- guard loop bodies with if condition
- reduce the inner deposition loop by one
- remove if condition before the operation `atomicAdd`

- [x] Test (all shapes 2/3D) KHI

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
**test case k20:**
- KHI electrons/positrons
- 4 GPUs (k20)
- particle drift in X direction
- global cells:
  - 3D - 192x512x12
  - 2D - 192x512
- SuperCellSize 8x8x4
- 2000 time steps
- 25 ppc/species
- used CUDA 8.0

speedup normalized to Esirkepov(dev)

| dimension | shape | Esirkepov(dev) [s] | Esirkepov(branch) [s] | speedup |
| --- | --: | --: | --: | --: |
3D | CIC | 298 | 173 | 1.7225
3D | TSC | 714 | 425 | 1.6800
3D | PCS | 1591 | 1025 | 1.5521
2D | CIC | 35 | 26 | 1.3461
2D | TSC | 56 | 42 | 1.3333
2D | PCS | 84 | 65 | 1.2923
2D | P4S | 234  |178  |1.3146

speedup normalized to EmZ

| dimension | shape | EmZ [s] | Esirkepov(branch) [s] | speedup |
| --- | --: | --: | --: | --: |
3D | CIC  | 151 |  173  | .8728
3D | TSC  | 371 |  425 |  .8729
3D | PCS |  916 |  1025  | .8936
2D | CIC |  25  | 26  | .9615
2D | TSC |  39 |  42  | .9285
2D | PCS  | 61 |  65 |  .9384
2D | P4S | 122  | 178  | .6853

**Comment by psychocoderHPC** [comment]
The heating and charge conservation for the  `KelvinHelmholtz` example is equal to the current dev for all shapes, 3D and 2D(X direction flow).

e.g. the TSC shape output:

charge_0009tsc

heat_0013tsc

**Comment by ax3l** [comment]
awesome!
Can you add the change in *local current conservation*, too please? (`plot_chargeConservation.py`)

note to others: documented in slides, we will put that optimization in the EmZ paper.

**Comment by psychocoderHPC** [comment]
ready for a review

**Comment by ax3l** [comment]
thanks, are the last plots `dev`, PR or identical in both cases?

File: src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov.hpp [hunk 1]
```diff
@@ -76,101 +66,127 @@ struct Esirkepov<T_ParticleShape, DIM3>
...
+            leaveCell,
+            cursorJ,
+            line,
+            cellSize.z()
+        );
     }
 
     /**
      * deposites current in z-direction
+     * \param leaveCell vector with information (for each direction) if the particle is leaving the cell
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov.hpp]: add: *\param leaveCell vector with information if the particle is leaving the cell (for each direction, __0 means stays in cell and 1 means leaves cell__)*

File: src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov2D.hpp [hunk 2]
```diff
@@ -120,84 +111,95 @@ struct Esirkepov<T_ParticleShape, DIM2>
...
+            leaveCell,
+            cursorJ,
+            line,
+            velocity.z()
+        );
     }
 
     /**
      * deposites current in z-direction
+     * \param leaveCell vector with information (for each direction) if the particle is leaving the cel
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov2D.hpp]: add: same information as above about 0 and 1 meaning :)

**Comment by ax3l** [comment]
let me just say once more: really great work, thank you and impressive! :fireworks: :sparkles:

File: src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov.hpp [hunk 3]
```diff
@@ -127,7 +128,8 @@ struct Esirkepov<T_ParticleShape, DIM3>
 
     /**
      * deposites current in z-direction
-     * \param leaveCell vector with information (for each direction) if the particle is leaving the cell
+     *
+     * \param leaveCell vector with information if the particle is leaving the cell (for each direction, 0 means stays in cell and 1 means leaves cell)
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov.hpp]:
```C++
* \param leaveCell vector with information if the particle is leaving the cell
*                  (for each direction, 0 means stays in cell and 1 means leaves cell)
```

File: src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov2D.hpp [hunk 4]
```diff
@@ -133,10 +133,12 @@ struct Esirkepov<T_ParticleShape, DIM2>
 
     /**
      * deposites current in z-direction
-     * \param leaveCell vector with information (for each direction) if the particle is leaving the cel
+     * \param leaveCell vector with information (for each direction) if the particle is leaving the cell
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/Esirkepov/Esirkepov2D.hpp]:
```C++
* \param leaveCell vector with information if the particle is leaving the cell
*                  (for each direction, 0 means stays in cell and 1 means leaves cell)
```

**Pull request merged**